File: frontend/src/api/captcha.ts
import api from './client'
import type { AnonCaptchaScene, CaptchaScene } from '../utils/captcha'

export interface CaptchaChallenge {
  challenge_id: string
  scene: CaptchaScene
  width: number
  height: number
  thumb_y: number
  thumb_width: number
  thumb_height: number
  image: string
  thumb: string
  expires_at: string
}

export interface CaptchaVerifyResult {
  scene: CaptchaScene
  captcha_token: string
}

export async function createCaptchaChallenge(scene: CaptchaScene) {
  const { data } = await api.post<CaptchaChallenge>('/captcha/challenges', { scene })
  return data
}

export interface TrajectoryPoint {
  x: number
  t: number
}

export async function verifyCaptchaChallenge(
  challengeId: string,
  x: number,
  y: number,
  trajectory: TrajectoryPoint[],
) {
  const { data } = await api.post<CaptchaVerifyResult>('/captcha/challenges/verify', {
    challenge_id: challengeId,
    x,
    y,
    trajectory,
  })
  return data
}

// ── 匿名端点（注册 / 登录，无需登录态）──────────────────────────

export interface AnonCaptchaChallenge {
  challenge_id: string
  scene: AnonCaptchaScene
  width: number
  height: number
  thumb_y: number
  thumb_width: number
  thumb_height: number
  image: string
  thumb: string
  expires_at: string
}

export interface AnonCaptchaVerifyResult {
  scene: AnonCaptchaScene
  captcha_token: string
}

export async function createAnonCaptchaChallenge(sessionId: string, scene: AnonCaptchaScene) {
  const { data } = await api.post<AnonCaptchaChallenge>('/captcha/anon/challenges', {
    session_id: sessionId,
    scene,
  })
  return data
}

export async function verifyAnonCaptchaChallenge(
  sessionId: string,
  challengeId: string,
  x: number,
  y: number,
  trajectory: TrajectoryPoint[],
) {
  const { data } = await api.post<AnonCaptchaVerifyResult>('/captcha/anon/challenges/verify', {
    session_id: sessionId,
    challenge_id: challengeId,
    x,
    y,
    trajectory,
  })
  return data
}

File: backend/app/schemas/captcha.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

CaptchaScene = Literal['view_worker_contact', 'chat_send', 'task_publish', 'task_accept']
AnonCaptchaScene = Literal['register', 'login']


class CaptchaChallengeCreate(BaseModel):
    scene: CaptchaScene


class CaptchaChallengeOut(BaseModel):
    challenge_id: str
    scene: CaptchaScene
    width: int
    height: int
    thumb_y: int
    thumb_width: int
    thumb_height: int
    image: str
    thumb: str
    expires_at: datetime


class TrajectoryPoint(BaseModel):
    x: float
    t: float


class CaptchaVerifyRequest(BaseModel):
    challenge_id: str = Field(min_length=8, max_length=64)
    x: int
    y: int
    trajectory: list[TrajectoryPoint] = Field(min_length=3, max_length=500)


class CaptchaVerifyOut(BaseModel):
    scene: CaptchaScene
    captcha_token: str


# ── 匿名场景（注册 / 登录）──────────────────────────────────────

class AnonCaptchaChallengeCreate(BaseModel):
    session_id: str = Field(min_length=8, max_length=64)
    scene: AnonCaptchaScene


class AnonCaptchaChallengeOut(BaseModel):
    challenge_id: str
    scene: AnonCaptchaScene
    width: int
    height: int
    thumb_y: int
    thumb_width: int
    thumb_height: int
    image: str
    thumb: str
    expires_at: datetime


class AnonCaptchaVerifyRequest(BaseModel):
    session_id: str = Field(min_length=8, max_length=64)
    challenge_id: str = Field(min_length=8, max_length=64)
    x: int
    y: int
    trajectory: list[TrajectoryPoint] = Field(min_length=3, max_length=500)


class AnonCaptchaVerifyOut(BaseModel):
    scene: AnonCaptchaScene
    captcha_token: str
